### PR TITLE
Introduce suppressReasons

### PR DESCRIPTION
**Pull request by BraisGabin**
First step to implement #3749. The next step is to use this for the suppression and the baseline.

### CODE REVIEW

File: detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/CodeSmell.kt [hunk 1]
```diff
@@ -10,7 +10,8 @@ package io.gitlab.arturbosch.detekt.api
 open class CodeSmell(
     final override val entity: Entity,
     final override val message: String,
-    final override val references: List<Entity> = emptyList()
+    final override val references: List<Entity> = emptyList(),
+    final override val suppress: List<String> = emptyList(),
```
Comment (cortinico) [inline detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/CodeSmell.kt:14]: nit: I think from the name is not immediately clear what this field is about.

I think it should be like `suppressions` or something more descriptive.

Comment (BraisGabin) [reply detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/CodeSmell.kt:14]: You are completely right. The name doesn't explain what it is. What do you think about `suppressReasons`? I'm completely fine with your proposal. I'm just thinking about other options.

Comment (cortinico) [reply detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/CodeSmell.kt:14]: Yeah that would work 👍

Comment (BraisGabin) [reply detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/CodeSmell.kt]: Rename done :)

**Review by cortinico — APPROVED** [review]

**Pull request merged**